### PR TITLE
Add OpenStack Swift scaler

### PR DESCRIPTION
**Pull request by pdominguite**
Signed-off-by: Pedro Dominguite p.dominguite@sidi.org.br

This PR adds a new scaler for the Swift OpenStack object-store.

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO)
- [x] Tests have been added
- [x] A PR is opened to update the documentation on https://github.com/kedacore/keda-docs (kedacore/keda-docs#342)
- [x] Changelog has been updated

We are committed to delivering end to end test for this scaler soon.

Related to #1342.

### CODE REVIEW

File: pkg/scaling/scale_handler.go
Comment (zroubalik) [inline pkg/scaling/scale_handler.go:467]:
```suggestion
	case "openstack-swift":
```

File: pkg/scalers/openstack_swift_scaler.go
Comment (zroubalik) [inline pkg/scalers/openstack_swift_scaler.go:361]: The generated `metricName` should be unique enough, that if there are multiple scalers of the same type in one ScaledObject, each metric name is still unique.
Just double checking, because I don't have enough expertise in Openstack Swift, I am not sure whether this is the case? 

Do you think that there's a sane scenario where we would like to have multiple swift triggers pointing to the same container? Or is that something, that doesn't make a sense?

File: pkg/scalers/openstack_swift_scaler.go
Comment (zroubalik) [inline pkg/scalers/openstack_swift_scaler.go:40]: For handling the http, could you please take a look on this PR: https://github.com/kedacore/keda/pull/1251

there's `kedautil.CreateHTTPClient(timeout time.Duration)` function.

**Review by zroubalik — COMMENTED** [review]
Thanks for the contribution!

Looking good, just some minor remarks

File: pkg/scaling/scale_handler.go
Comment (pdominguite) [reply pkg/scaling/scale_handler.go:467]: Constructor and other helper functions were named `NewSwiftScaler` or `parseSwiftMetadata`. Would it be the case to rename them to `NewOpenStackSwiftScaler` and `parseOpenStackSwiftMetadata`, for example?

Comment (zroubalik) [reply pkg/scaling/scale_handler.go:467]: yeah

File: pkg/scalers/openstack_swift_scaler.go
Comment (pdominguite) [reply pkg/scalers/openstack_swift_scaler.go:361]: OpenStack Swift should work similarly to Azure Blob Storage or any other object-store service.

We have identified one possible scenario where it could be possible to have multiple swift triggers pointing to the same container, though: when one wants to scale based on the number of objects inside different folders inside the same container. In this case, delimiter and prefixes can be used. In order to set more than one prefix, new triggers for the same container in the same ScaledObject must be created.

Thus, we have modified `metricName` to contain the `containerName` + `objectPrefix`, when the latter is provided. That should be enough to cover this scenario.

File: pkg/scalers/openstack_swift_scaler.go
Comment (pdominguite) [reply pkg/scalers/openstack_swift_scaler.go:40]: Done by 07fc5e4.

File: pkg/scaling/scale_handler.go
Comment (pdominguite) [reply pkg/scaling/scale_handler.go]: Both 0313be5 and e12e140 applied the suggested changes 😃

File: pkg/scalers/openstack_swift_scaler.go
Comment (zroubalik) [reply pkg/scalers/openstack_swift_scaler.go]: Sounds good, thanks!

**Review by zroubalik — APPROVED** [review]
LGTM, thanks @pdominguite!

What is your plan for e2e tests? We would like to see one for this scaler.

**Comment by pdominguite** [comment]
Thanks for all feedbacks @zroubalik! We are very happy to contribute to Keda 😄 

Regarding e2e tests, we plan to submit them in a separate PR by the end of this month/beginning of February. Is that ok? End-to-end tests are a must-have to add this contribution or the swift scaler can be merged and tests provided later on?

File: pkg/scalers/openstack/keystone_authentication.go
Comment (ahmelsayed) [inline pkg/scalers/openstack/keystone_authentication.go:151]: it seems to me that you don't need to make `passAuth`, `passAuth.Properties`, `passAuth.Properties.Identity`, `passAuth.Properties.Identity.Password`, `passAuth.Properties.Identity.Password.User`, or `passAuth.Properties.Scope` references, and then you'll get a default object for all of them just by doing 

```go
passAuth := KeystoneAuthMetadata{}
```

right?

File: pkg/scalers/openstack/keystone_authentication.go
Comment (ahmelsayed) [inline pkg/scalers/openstack/keystone_authentication.go:98]: nit: since this is an error case, maybe handle the err in case the stream is terminated from the server early.

**Comment by ahmelsayed** [comment]
I think an e2e test can come in a separate PR to not block this one on it.

File: pkg/scalers/openstack/keystone_authentication.go
Comment (pdominguite) [reply pkg/scalers/openstack/keystone_authentication.go:155]: As far as I know (and tried), since `KeystoneAuthMetadata` is a struct that has other structs inside that are pointers, they need to be initialized somehow first. Thus, 

```go
passAuth := KeystoneAuthMetadata{}
```

is not sufficient to get a proper `KeystoneAuthMetadata` default object. Assigning values to it without manually making those references will result in: 
```
runtime error: invalid memory address or nil pointer dereference
```

Maybe there is a more elegant way to do or another way to initialize these "inner structs", but my limited experience in Go does not allow me to find it out 😅

File: pkg/scalers/openstack/keystone_authentication.go
Comment (pdominguite) [reply pkg/scalers/openstack/keystone_authentication.go]: Sorry, I missed this error handling. c6d9f14 added the basic handler.

**Comment by pdominguite** [comment]
Should I interactive rebase this branch to fixup commits only to "Add new OpenStack Swift scaler"?

File: pkg/scalers/openstack/keystone_authentication.go
Comment (ahmelsayed) [reply pkg/scalers/openstack/keystone_authentication.go:155]: I guess what I meant is maybe non of these pointers need to be a pointer since you never expect them to actually be nil, right? So `KeystoreAuthMetadata` would be 

```diff
type KeystoneAuthMetadata struct {
	AuthURL    string       `json:"-"`
	AuthToken  string       `json:"-"`
	HTTPClient *http.Client `json:"-"`
-	Properties *authProps   `json:"auth"`
+	Properties authProps   `json:"auth"`
}
```

and in turn `authProps` would be 

```diff
type authProps struct {
-	Identity *identityProps `json:"identity"`
+	Identity identityProps `json:"identity"`
-	Scope    *scopeProps    `json:"scope,omitempty"`
+	Scope    scopeProps    `json:"scope,omitempty"`
}
```

and so on. 

I think the only difference is in the `json.Marshal()` behavior where a nil is an empty object but any non-nil object is not empty, but I don't think it makes a difference really. Just thought it would make the code a bit easier to read since you're already initializing all the pointers with new() anyway. No worries if it gives you trouble.

Comment (pdominguite) [reply pkg/scalers/openstack/keystone_authentication.go:155]: Exactly! They do not need to be pointers. The only reason we did that way was because of `json.Marshal()` as you noticed. Since we use these structs to build a JSON payload for our token request, some of them need to be omitted (and thus be pointers, and then be `nil`) in order to make it work properly.

Also, we plan to submit new scalers for OpenStack projects soon. As they come, we can study better ways to do it and, eventually, updated the way we originally structured this.

That is nice feedback, anyway @ahmelsayed. Thanks! We'll keep an eye on it 😄

**Pull request merged**
